Sources/SwiftProtobuf/TextDecoder.swift

```diff
@@ -0,0 +1,211 @@
+                if key.hasPrefix("[") {
+                    // Extension key
+                    if let n = scanner.extensions?.fieldNumberForProto(messageType: M.self, protoFieldName: key) {
+                        protoFieldNumber = n
+                    } else {
+                        throw DecodingError.unknownField
+                    }
+                } else {
+                    // Regular key; look it up on the message
+                    if let n = nameProviding._protobuf_fieldNames.fieldNumber(forProtoName: key) {
+                        protoFieldNumber = n
+                    } else {
+                        throw DecodingError.unknownField
+                    }
+                }
+                var fieldDecoder: FieldDecoder = TextFieldDecoder(scanner: scanner)
+                try message.decodeField(setter: &fieldDecoder, protoFieldNumber: protoFieldNumber)
+            default:
+                if terminator != nil && terminator == token {
+                    return
+                }
+                throw DecodingError.malformedText
+            }
+            try scanner.skipOptionalSeparator()
+        }
+        if terminator == nil {
+            return
+        } else {
+            throw DecodingError.truncatedInput
+        }
+    }
+}
+
+struct TextFieldDecoder: FieldDecoder {
+    var rejectConflictingOneof: Bool {return true}
+    var scanner: TextScanner
+
+    mutating func decodeExtensionField(values: inout ExtensionFieldValueSet, messageType: Message.Type, protoFieldNumber: Int) throws {
+        if let ext = scanner.extensions?[messageType, protoFieldNumber] {
+            var mutableSetter: FieldDecoder = self
+            var fieldValue = values[protoFieldNumber] ?? ext.newField()
+            try fieldValue.decodeField(setter: &mutableSetter)
+            values[protoFieldNumber] = fieldValue
+        }
+    }
+
+    mutating func decodeSingularField<S: FieldType>(fieldType: S.Type, value: inout S.BaseType?) throws {
+        try scanner.skipRequired(token: .colon)
+        try S.setFromText(scanner: scanner, value: &value)
+    }
+
+    mutating func decodeRepeatedField<S: FieldType>(fieldType: S.Type, value: inout [S.BaseType]) throws {
+        try scanner.skipRequired(token: .colon)
+        if try scanner.skipOptional(token: .beginArray) {
+            var firstItem = true
+            while true {
+                if try scanner.skipOptional(token: .endArray) {
+                    return
+                }
+                if firstItem {
+                    firstItem = false
+                } else {
+                    try scanner.skipRequired(token: .comma)
+                }
+                try S.setFromText(scanner: scanner, value: &value)
+            }
+        } else {
+            try S.setFromText(scanner: scanner, value: &value)
+        }
+    }
+
+    mutating func decodePackedField<S: FieldType>(fieldType: S.Type, value: inout [S.BaseType]) throws {
```

Comment:
Yes:  this is how the generated code communicates the type that was declared in the proto file.  Since proto files can have `packed` fields (it's even the default in proto3 for many types), all decoders must implement this.